README.md
pyxtrlock
=========

``pyxtrlock`` is a rewrite of Ian Jackson's great ``xtrlock`` program using
modern libraries, most importantly the obsolete direct passwd/shadow
authentication has been replaced by today's
[PAM](https://en.wikipedia.org/wiki/Pluggable_authentication_module) authentication
mechanism, hence it also works on Fedora. Also, it's mostly written using
[XCB](http://xcb.freedesktop.org/) instead of Xlib, although some Xlib/XCB
interaction is still necessary. As soon as XCB can provide the required
services of Xlib, the remaining code will be ported to XCB.

Since pyxtrlock uses PAM authentication, it can be run as the normal user and
doesn't need special privileges.

Description
-----------
pyxtrlock, like its predecessor, is a very minimal X display lock program. While
pyxtrlock is running, it doesn't obscure the screen, only the mouse and keyboard
are grabbed and the mouse cursor becomes a padlock. Output displayed by X
programs, and windows put up by new X clients, continue to be visible, and any
new output is displayed normally.

In good Unix tradition, pyxtrlock provides no visual feedback during password
entry. You simply type your password at it, followed by Enter or Newline.
Pressing Backspace or Delete erases one character of a password partially typed;
pressing Escape or Clear clears anything that has been entered.

Like xtrlock, pyxtrlock will ignore further keystrokes until a timeout has
expired after too many attempts have been made in too short time.

Installation and Usage
----------------------
Install [python3-simplepam](https://github.com/leonnnn/python3-simplepam) for
dependencies:

    $ git clone git://github.com/leonnnn/python3-simplepam.git
    $ cd python3-simplepam
    $ sudo python3 setup.py install

Install pyxdg, which is available as python3-pyxdg or similar in most Linux
distributions.

Clone and install pyxtrlock:

    $ git clone git://github.com/leonnnn/pyxtrlock.git
    $ cd pyxtrlock
    $ sudo python3 setup.py install

Once this is done, you should be able to simply lock your display by running

    $ pyxtrlock

If you would like to automatically lock your screen after some idle time,
we recommend the ``xautolock`` tool. Just add something like

    xautolock -locker pyxtrlock -time 5

to your X autostart file to lock the screen with ``pyxtrlock`` after 5
minutes idle time. ``xautolock`` has many other useful features, see
its documentation. Most distributions provide an ``xautolock`` package
with a man page.

An modern alternative to ``xautolock`` is ``xss-lock`` which
integrates with ``systemd-logind`` and manages locking on idleness and
lid close events. ``pyxtrlock`` supports the ``XSS_SLEEP_LOCK_FD``
protocol used by ``xss-lock -l`` to delay system standby until the
screen is locked, the recommended way to start ``pyxtrlock`` with
``xss-lock`` is:

    xss-lock -l -- pyxtrlock

to explicitly lock the screen use:

    loginctl lock-session

Staying up-to-date
------------------
As pyxtrlock is a security tool, it is important to stay up-to-date with
security updates. We take security seriously and try to handle any
vulnerabilities quickly. However, our efforts are useless if the users
aren’t notified that updates are available, so if you use pyxtrlock, we
urge you to subscribe to
[the pyxtrlock mailing list](http://lists.zombofant.net/mailman/listinfo/pyxtrlock).
This list is likely very low traffic and will ensure you get
notifications of security updates in time.

We also appreciate any feedback you have regarding pyxtrlock on this
mailing list.

Bugs & Limitations
------------------
Pyxtrlock can only securely lock X11 sessions. Running pyxtrlock in other
session types, such as Wayland, is insecure and not supported. Pyxtrlock tries
to detect Wayland sessions and abort with an error message.

Pyxtrlock does not prevent a user from switching to a virtual
terminal, so be advised to always log out from your terminals.

Additional input devices other than the keyboard and mouse are not disabled.

The length of the password is limited to 100 KiB to prevent memory
exhaustion attacks. This limit can only be adapted in the source code.

The width and height of the cursor bitmaps is limited to 512 pixels
(primarily to protect the user from faulty cursor files). This limit
can be only adapted in the source code.

Please report any new bugs you may find to our
[Github issue tracker](https://github.com/leonnnn/pyxtrlock/issues).

Configuration
-------------
The padlock icon can be changed. While the default lock is stored in
the source code, an alternative lock can be stored in one of the the
xdg data paths for pyxtrlock.

The user configured lock is stored as a json file containing the
necessary information. ``bin/make_lock.py`` is a tool for generating
cursors from image files. See ``doc/make_lock.txt`` for the full
documentation of the tool.

Note, that even though loading json does not allow arbitrary code
execution and the cursor data is checked for consistency, cursor files
should be created on your machine and should not be installed from
untrusted sources.

Requirements
------------
* [python3-simplepam](https://github.com/leonnnn/python3-simplepam)
* [pyxdg](http://freedesktop.org/Software/pyxdg)
* Python ≥ 3.0
* libxcb
* libxcb-image
* libX11 ≥ 1.4, or libX11 ≥ 1.2 compiled with XCB backend

These requirements are met at least on
* Debian wheezy and sid, and probably on Debian squeeze (untested; please report your experience)
* Ubuntu ≥ 11.10, and probably 10.04 (untested; please report your experience)
* Fedora ≥ 16

Authors
-------
* Leon Weber <leon@leonweber.de>
* Sebastian Riese <s.riese@zombofant.net>

pyxtrlock has been inspired by
[Ian Jacksons](http://www.chiark.greenend.org.uk/~ijackson/)'s brilliant
``xtrlock`` program and uses many ideas and techniques from the xtrlock
source code. Also, the lock icon has been copied from xtrlock.

License
-------
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, either version 3 of the License, or
(at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <http://www.gnu.org/licenses/>.


CHANGELOG
Release 0.4:
• Security: Refuse to run if wayland session is detected. Running in a wayland
  is insecure.

Release 0.3:
• Enhancement: Make pyxtrlock conform to the XDG specification. Patch by
  Cristian Ciupitu.
• Enhancement: Ship script to create an empty lock symbol. Patch by Alexander
  Klink.
• Enhancement: Retry grabbing the keyboard as well as the mouse to
  prevent issues when starting pyxtrlock from a keyboard shortcut.
  Patch by Nils Ballmann.
• Enhancement: Support the XSS_SLEEP_LOCK_FD protocol implemented by
  xss-lock with the option -l to inhibit standby/hibernate/... until
  the the screen is locked.

Release 0.2 :
• Security: Check correctly for the result of the
  xcb_grab_{pointer,keyboard} commands (CVE-2013-4427).
• Security: Limit length of buffered password to prevent memory exhaustion
  (this is a real concern when attacked with custom hardware which
  simulates most rapid keystrokes)
• Security: Fix several memory leaks
• Enhancement: Report missing libraries when loading via ctypes.
• Enhancement: Provide ability and tools to use custom lock images as cursor

Release 0.1:
• [#8] Security: Fixed a typo that could in some circumstances lead to a
       crash after multiple failed authentication attempts. Thanks,
       Paul Lhussiez (CVE-2013-4426).


pyxtrlock/xcb.py
from ctypes import *
from pyxtrlock.utils import check_and_load_library


class XCBError(Exception):
    """Raised on XCBErrors.

    This encapsulates the underlying xcb error object. Note that the
    xcb error is freed when this object is garbage collected. So you
    should never copy the xcb_error property.
    """
    def __init__(self, xcb_error):
        self.xcb_error = xcb_error

    def __del__(self):
        free(self.xcb_error)


class Connection(Structure):
    pass


class Setup(Structure):
    pass


Window = c_uint32
Colormap = c_uint32
VisualID = c_uint32


class Screen(Structure):
    _fields_ = [
        ("root", Window),
        ("default_colormap", Colormap),
        ("white_pixel", c_uint32),
        ("black_pixel", c_uint32),
        ("current_input_masks", c_uint32),
        ("width_in_pixels", c_uint16),
        ("height_in_pixels", c_uint16),
        ("width_in_millimeters", c_uint16),
        ("height_in_millimeters", c_uint16),
        ("min_installed_maps", c_uint16),
        ("max_installed_maps", c_uint16),
        ("root_visual", VisualID),
        ("backing_stores", c_uint8),
        ("save_unders", c_uint8),
        ("root_depth", c_uint8),
        ("allowed_depths_len", c_uint8)
    ]


class ScreenIterator(Structure):
    _fields_ = [
        ("data", POINTER(Screen)),
        ("rem", c_int),
        ("index", c_int)
    ]


class GetWindowAttributesReply(Structure):
    _fields_ = [
        ("response_type", c_uint8),
        ("backing_store", c_uint8),
        ("sequence", c_uint16),
        ("length", c_uint32),
        ("visual", VisualID),
        ("_class", c_uint16),
        ("bit_gravity", c_uint8),
        ("win_gravity", c_uint8),
        ("backing_planes", c_uint32),
        ("backing_pixel", c_uint32),
        ("save_under", c_uint8),
        ("map_is_installed", c_uint8),
        ("map_state", c_uint8),
        ("override_redirect", c_uint8),
        ("colormap", Colormap),
        ("all_event_masks", c_uint32),
        ("your_event_masks", c_uint32),
        ("do_not_propagate_mask", c_uint16),
        ("pad0", c_uint8 * 2)
    ]


class Cookie(Structure):
    _fields_ = [
        ("sequence", c_uint)
    ]


VoidCookie = Cookie
AllocNamedColorCookie = Cookie
AllocColorCookie = Cookie
GrabKeyboardCookie = Cookie
GrabPointerCookie = Cookie


# TODO?
class ImageFormat(c_int):
    pass


class ImageOrder(c_int):
    pass


class AllocNamedColorReply(Structure):
    _fields_ = [
        ("response_type", c_uint8),
        ("pad0", c_uint8),
        ("sequence", c_uint16),
        ("length", c_uint32),
        ("pixel", c_uint32),
        ("exact_red", c_uint16),
        ("exact_green", c_uint16),
        ("exact_blue", c_uint16),
        ("visual_red", c_uint16),
        ("visual_green", c_uint16),
        ("visual_blue", c_uint16)
    ]


class AllocColorReply(Structure):
    _fields_ = [
        ("response_type", c_uint8),
        ("pad0", c_uint8),
        ("sequence", c_uint16),
        ("length", c_uint32),
        ("red", c_uint16),
        ("green", c_uint16),
        ("blue", c_uint16),
        ("pad1", c_uint8 * 2),
        ("pixel", c_uint32),
    ]


class GenericError(Structure):
    _fields_ = [
        ("response_type", c_uint8),
        ("error_code", c_uint8),
        ("sequence", c_uint16),
        ("resource_id", c_uint32),
        ("minor_code", c_uint16),
        ("major_code", c_uint8),
        ("pad0", c_uint8),
        ("pad", c_uint32 * 5),
        ("full_sequence", c_uint32)
    ]

    def __str__(self):
        return '{}({})'.format(self.__class__.__name__,
                               ','.join(field + '=' + str(getattr(self, field))
                                        for field, _ in self._fields_))


class GenericID(c_uint32):
    pass


class GrabReply(Structure):
    _fields_ = [
        ("response_type", c_uint8),
        ("status", c_uint8),
        ("sequence", c_uint16),
        ("length", c_uint32)
    ]

    def __str__(self):
        return '{}({})'.format(self.__class__.__name__,
                               ','.join(str(getattr(self, field))
                                        for field, _ in self._fields_))


class GenericEvent(Structure):
    _fields_ = [
        ("response_type", c_uint8),
        ("pad0", c_uint8),
        ("sequence", c_uint16),
        ("pad", c_uint32 * 7),
        ("full_sequence", c_uint32)
    ]


Keycode = c_uint8
Timestamp = c_uint32


class KeyPressEvent(Structure):
    _fields_ = [
        ("response_type", c_uint8),
        ("detail", Keycode),
        ("sequence", c_uint16),
        ("time", Timestamp),
        ("root", Window),
        ("event", Window),
        ("child", Window),
        ("root_x", c_int16),
        ("root_y", c_int16),
        ("event_x", c_int16),
        ("event_y", c_int16),
        ("state", c_uint16),
        ("same_screen", c_uint8),
        ("pad0", c_uint8)
    ]


GrabKeyboardReply = GrabReply
GrabPointerReply = GrabReply


Pixmap = GenericID
Cursor = GenericID

COPY_FROM_PARENT = 0
WINDOW_CLASS_INPUT_ONLY = 2
CW_OVERRIDE_REDIRECT = 512
CW_EVENT_MASK = 2048

EVENT_MASK_KEY_PRESS = 1
EVENT_MASK_KEY_RELEASE = 2

CURRENT_TIME = 0
GRAB_MODE_ASYNC = 1

WINDOW_NONE = 0

KEY_PRESS = 2

libxcb = check_and_load_library('xcb')
libxcb_image = check_and_load_library('xcb-image')
libc = check_and_load_library('c')

connect = libxcb.xcb_connect
connect.argtypes = [c_char_p, POINTER(c_int)]
connect.restype = POINTER(Connection)

get_setup = libxcb.xcb_get_setup
get_setup.argtypes = [POINTER(Connection)]
get_setup.restype = POINTER(Setup)

setup_roots_iterator = libxcb.xcb_setup_roots_iterator
setup_roots_iterator.argtypes = [POINTER(Setup)]
setup_roots_iterator.restype = ScreenIterator

screen_next = libxcb.xcb_screen_next
screen_next.argtypes = [POINTER(ScreenIterator)]
screen_next.restype = None

generate_id = libxcb.xcb_generate_id
generate_id.argtypes = [POINTER(Connection)]
generate_id.restype = GenericID

create_window = libxcb.xcb_create_window
create_window.argtypes = [
    POINTER(Connection),    # connection
    c_uint8,    # depth
    Window,     # wid
    Window,     # parent
    c_int16,    # x
    c_int16,    # y
    c_uint16,   # width
    c_uint16,   # height
    c_uint16,   # border_width
    c_uint16,   # _class
    VisualID,   # visual
    c_uint32,   # value_mask
    POINTER(c_uint32)   # value_lis
]
create_window.restype = VoidCookie

alloc_named_color = libxcb.xcb_alloc_named_color
alloc_named_color.argtypes = [
    POINTER(Connection),    # connection
    Colormap,   # cmap
    c_uint16,   # name len
    c_char_p    # name
]
alloc_named_color.restype = AllocNamedColorCookie

alloc_color = libxcb.xcb_alloc_color
alloc_color.argtypes = [
    POINTER(Connection),    # connection
    Colormap,   # cmap
    c_uint16,   # r
    c_uint16,   # g
    c_uint16    # b
]
alloc_color.restype = AllocColorCookie

alloc_named_color_reply = libxcb.xcb_alloc_named_color_reply
alloc_named_color_reply.argtypes = [
    POINTER(Connection),    # connection
    AllocNamedColorCookie,  # cookie
    POINTER(POINTER(GenericError))  # e
]
alloc_named_color_reply.restype = POINTER(AllocNamedColorReply)

alloc_color_reply = libxcb.xcb_alloc_color_reply
alloc_color_reply.argtypes = [
    POINTER(Connection),    # connection
    AllocColorCookie,  # cookie
    POINTER(POINTER(GenericError))  # e
]
alloc_color_reply.restype = POINTER(AllocColorReply)


def alloc_named_color_sync(conn, colormap, color_string):
    """Synchronously allocate a named color

    Wrapper function for xcb_alloc_named_color and alloc_named_color_reply.

    Raises ``XCBError`` on errors.
    """
    if isinstance(color_string, str):
        color_string = color_string.encode('us-ascii')

    cookie = alloc_named_color(conn, colormap, len(color_string),
                               color_string)
    error_p = POINTER(GenericError)()
    res = alloc_named_color_reply(conn, cookie, byref(error_p))
    if error_p:
        raise XCBError(error_p.contents)

    ret = (res.contents.visual_red, res.contents.visual_green,
           res.contents.visual_blue)
    free(res)
    return ret


def alloc_color_sync(conn, colormap, r, g, b):
    """Synchronously allocate a color

    Wrapper function for xcb_alloc_color and alloc_color_reply.

    The (r, g, b) triple is in the range 0 to 255 (as opposed to
    the X protocol using the 0 to 2^16-1 range).

    Raises ``XCBError`` on xcb errors and value errors for invalid
    values of r, g, b.
    """
    if r < 0 or g < 0 or b < 0:
        raise ValueError("color component out of range")
    if r > 255 or g > 255 or b > 255:
        raise ValueError("color component our of range")

    r <<= 8; g <<= 8; b <<= 8

    cookie = alloc_color(conn, colormap, r, g, b)
    error_p = POINTER(GenericError)()
    res = alloc_color_reply(conn, cookie, byref(error_p))
    if error_p:
        raise XCBERror(error_p.contents)

    ret = (res.contents.red, res.contents.green, res.contents.blue)
    free(res)
    return ret


request_check = libxcb.xcb_request_check
request_check.argtypes = [POINTER(Connection), VoidCookie]
request_check.restype = POINTER(GenericError)

create_cursor_checked = libxcb.xcb_create_cursor_checked
create_cursor_checked.argtypes = [
    POINTER(Connection),    # connection
    Cursor,     # cursor
    Pixmap,     # source
    Pixmap,     # mask
    c_uint16,   # fore_red
    c_uint16,   # fore_green
    c_uint16,   # fore_blue
    c_uint16,   # back_red
    c_uint16,   # back_green
    c_uint16,   # back_blue
    c_uint16,   # x
    c_uint16    # y
]
create_cursor_checked.restype = VoidCookie


def create_cursor_sync(conn, source, mask, fg, bg, x, y):
    """
    Synchronously create a cursor, including error checks.

    Raises ``XCBError`` on failure. Otherwise returns ``Cursor``.
    """
    cursor = generate_id(conn)
    cookie = create_cursor_checked(conn, cursor, source, mask,
                                   fg[0], fg[1], fg[2], bg[0],
                                   bg[1], bg[2], x, y)
    error = request_check(conn, cookie)
    if error:
        raise XCBError(error.contents)

    return cursor


map_window = libxcb.xcb_map_window
map_window.argtypes = [POINTER(Connection), Window]
map_window.restype = VoidCookie


flush = libxcb.xcb_flush
flush.argtypes = [POINTER(Connection)]
flush.restype = c_int


grab_keyboard = libxcb.xcb_grab_keyboard
grab_keyboard.argtypes = [
    POINTER(Connection),    # connection
    c_uint8,    # owner_events
    Window,     # grab_window
    Timestamp,  # time
    c_uint8,    # pointer_mode
    c_uint8     # keyboard_mode
]
grab_keyboard.restype = GrabKeyboardCookie


grab_keyboard_reply = libxcb.xcb_grab_keyboard_reply
grab_keyboard_reply.argtypes = [
    POINTER(Connection),    # connection
    GrabKeyboardCookie,     # cookie,
    POINTER(POINTER(GenericError))  # e
]
grab_keyboard_reply.restype = POINTER(GrabKeyboardReply)


def grab_keyboard_sync(conn, owner_events, grab_window, time, ptr_mode,
                       kbd_mode):
    """
    Synchronously grab the keyboard.

    Wrapper function for grab_pointer and grab_pointer_reply. Returns
    the status field from the reply. Raises ``XCBError`` on error.
    """
    owner_events = 1 if owner_events else 0

    cookie = grab_keyboard(conn, owner_events, grab_window, time, ptr_mode,
                           kbd_mode)
    error_p = POINTER(GenericError)()
    kbd_grab = grab_keyboard_reply(conn, cookie, byref(error_p))

    if error_p:
        raise XCBError(error_p.contents)
    status = kbd_grab.contents.status
    free(kbd_grab)
    return status


grab_pointer = libxcb.xcb_grab_pointer
grab_pointer.argtypes = [
    POINTER(Connection),    # connection
    c_uint8,    # owner_events
    Window,     # grab_window
    c_uint16,   # event_mask
    c_uint8,    # pointer_mode
    c_uint8,    # keyboard_mode
    Window,     # confine_to
    Cursor,     # cursor
    Timestamp  # time
]
grab_pointer.restype = GrabPointerCookie

grab_pointer_reply = libxcb.xcb_grab_pointer_reply
grab_pointer_reply.argtypes = [
    POINTER(Connection),    # connection
    GrabPointerCookie,     # cookie,
    POINTER(POINTER(GenericError))  # e
]
grab_pointer_reply.restype = POINTER(GrabPointerReply)


# constants to interpret grab results
GrabSuccess = 0
AlreadyGrabbed = 1
GrabInvalidTime = 2
GrabNotViewable = 3
GrabFrozen = 4


def grab_pointer_sync(conn, owner_events, window, event_mask, ptr_mode,
                      kbd_mode, confine_to, cursor, timestamp):
    """
    Synchronously grab the pointing device.

    Wrapper function for ``grab_pointer`` and ``grab_pointer_reply``.
    Raises ``XCBError`` on error. Otherwise returns the result of
    ``grab_pointer_reply``
    """
    owner_events = 1 if owner_events else 0
    cookie = grab_pointer(conn, owner_events, window, event_mask, ptr_mode,
                          kbd_mode, confine_to, cursor, timestamp)
    error_p = POINTER(GenericError)()
    ptr_grab = grab_pointer_reply(conn, cookie, byref(error_p))
    if error_p:
        raise XCBError(error_p.contents)
    status = ptr_grab.contents.status
    free(ptr_grab)
    return status


wait_for_event_ = libxcb.xcb_wait_for_event
wait_for_event_.argtypes = [POINTER(Connection)]
wait_for_event_.restype = POINTER(GenericEvent)


free = libc.free
free.argtypes = [c_void_p]
free.restype = None


class FreeWrapper(object):

    def __init__(self, pointer):
        self.pointer = pointer

    def __enter__(self):
        return self.pointer

    def __exit__(self, etype, evalue, traceback):
        free(self.pointer)


def wait_for_event(conn):
    return FreeWrapper(wait_for_event_(conn))


connection_has_error = libxcb.xcb_connection_has_error
connection_has_error.restype = c_int
connection_has_error.argtypes = [
    POINTER(Connection),
]


# xcb_image
image_create_pixmap_from_bitmap_data = \
    libxcb_image.xcb_create_pixmap_from_bitmap_data
image_create_pixmap_from_bitmap_data.restype = Pixmap
image_create_pixmap_from_bitmap_data.argtypes = [
    POINTER(Connection),     # connection
    Window,    # drawable
    c_char_p,    # data
    c_uint32,   # width
    c_uint32,   # height
    c_uint32,   # depth
    c_uint32,   # fg
    c_uint32,   # bg
    POINTER(c_uint8)   # XXX graphics context
]
